File: server/src/collectors/waybackJp.js
/**
 * Wayback Machine CDX API — historical snapshots of JP gov/corp sites.
 *
 * Surfaces: deleted advisories, removed personnel pages, old API endpoints,
 * and historical reorgs. Useful baseline for diffing what JP orgs *used to*
 * publish vs what they hide today.
 *
 * Endpoint:
 *   GET https://web.archive.org/cdx/search/cdx
 *     ?url=<host/prefix>&matchType=prefix&limit=100&output=json
 *
 * Free, no auth. Override targets with WAYBACK_TARGETS=host1,host2,...
 */

import { intelEnvelope, intelUid } from '../utils/intelHelpers.js';

const SOURCE_ID = 'wayback-jp';
const BASE = 'https://web.archive.org/cdx/search/cdx';
const TIMEOUT_MS = 35000;

const DEFAULT_TARGETS = (process.env.WAYBACK_TARGETS || [
  'kantei.go.jp',
  'mod.go.jp',
  'mofa.go.jp',
  'meti.go.jp',
  'soumu.go.jp',
  'jpcert.or.jp',
  'ipa.go.jp',
  'nisc.go.jp',
  'pmda.go.jp',
  'mhlw.go.jp',
].join(',')).split(',').map((s) => s.trim()).filter(Boolean);

const PER_TARGET_LIMIT = parseInt(process.env.WAYBACK_PER_TARGET_LIMIT || '20', 10);

async function fetchOne(host) {
  const params = new URLSearchParams({
    url: host,
    matchType: 'prefix',
    limit: String(PER_TARGET_LIMIT),
    output: 'json',
    fl: 'timestamp,original,statuscode,mimetype,digest',
    filter: 'statuscode:200',
    collapse: 'digest',
  });

  try {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), TIMEOUT_MS);
    const res = await fetch(`${BASE}?${params}`, {
      signal: controller.signal,
      headers: { accept: 'application/json' },
    });
    clearTimeout(timer);
    if (!res.ok) return [];
    const rows = await res.json();
    if (!Array.isArray(rows) || rows.length < 2) return [];
    const [header, ...data] = rows;
    return data.map((row) => {
      const obj = {};
      header.forEach((k, i) => { obj[k] = row[i]; });
      obj._target = host;
      return obj;
    });
  } catch {
    return [];
  }
}

export default async function collectWaybackJp() {
  const all = await Promise.all(DEFAULT_TARGETS.map(fetchOne));
  const flat = all.flat();

  const items = flat.map((row, i) => {
    const replay = row.timestamp && row.original
      ? `https://web.archive.org/web/${row.timestamp}/${row.original}`
      : null;
    let publishedIso = null;
    if (row.timestamp && /^\d{14}$/.test(row.timestamp)) {
      const t = row.timestamp;
      publishedIso = `${t.slice(0,4)}-${t.slice(4,6)}-${t.slice(6,8)}T${t.slice(8,10)}:${t.slice(10,12)}:${t.slice(12,14)}Z`;
    }
    return {
      uid: intelUid(SOURCE_ID, row.digest, `${row._target}|${row.timestamp}|${i}`),
      title: row.original || row._target,
      summary: `${row._target} · ${row.mimetype || 'unknown mime'} · ${row.statuscode || '?'}`,
      link: replay,
      language: 'ja',
      published_at: publishedIso,
      tags: ['wayback', `host:${row._target}`],
      properties: {
        target: row._target,
        timestamp: row.timestamp || null,
        original: row.original || null,
        mimetype: row.mimetype || null,
        statuscode: row.statuscode || null,
        digest: row.digest || null,
      },
    };
  });

  return intelEnvelope({
    sourceId: SOURCE_ID,
    items,
    description: 'Wayback Machine CDX snapshots for JP government / cyber agency hosts',
    extraMeta: {
      targets: DEFAULT_TARGETS,
      env_hint: 'Override targets via WAYBACK_TARGETS=host1,host2,...; tune WAYBACK_PER_TARGET_LIMIT',
    },
  });
}

File: native/collectors/cyber/sources/wayback_jp.c
/* collectors/cyber/sources/wayback_jp.c
 * Port of server/src/collectors/waybackJp.js (intelEnvelope).
 * Wayback CDX (output=json → [header,...rows]) for 10 fixed JP gov/cyber
 * hosts, one GET per host, flattened. uid = wayback-jp|<digest ||
 * <target>|<timestamp>|<flatIdx>>. WAYBACK_TARGETS env override not ported. */
#include "../../../source.h"
#include "../../../lib/feedlib.h"
#include "../../../third_party/cJSON.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static const char *TARGETS[] = {
  "kantei.go.jp", "mod.go.jp", "mofa.go.jp", "meti.go.jp", "soumu.go.jp",
  "jpcert.or.jp", "ipa.go.jp", "nisc.go.jp", "pmda.go.jp", "mhlw.go.jp",
};
#define NT ((int)(sizeof(TARGETS) / sizeof(TARGETS[0])))

/* obj[col] as a JS string-or-NULL (CDX gives strings; '' is falsy). */
static const char *cell(cJSON *row, int idx) {
  cJSON *v = cJSON_GetArrayItem(row, idx);
  if (!v) return NULL;
  if (cJSON_IsString(v)) return v->valuestring[0] ? v->valuestring : NULL;
  return NULL;
}

static int is_ts14(const char *s) {
  if (!s) return 0;
  int i = 0;
  for (; s[i]; i++) if (s[i] < '0' || s[i] > '9') return 0;
  return i == 14;
}

static int run(const source_ctx *ctx, intel_sink *sink) {
  const char *hdrs[] = { "accept: application/json", NULL };
  int n = 0;
  int flat = 0;                                     /* .flat() index */
  for (int ti = 0; ti < NT; ti++) {
    const char *host = TARGETS[ti];
    char url[384];
    snprintf(url, sizeof url,
      "https://web.archive.org/cdx/search/cdx?url=%s&matchType=prefix"
      "&limit=20&output=json&fl=timestamp,original,statuscode,mimetype,digest"
      "&filter=statuscode:200&collapse=digest", host);
    cJSON *rows = feed_get_json_h(ctx->http, url, hdrs, 35000);
    if (!cJSON_IsArray(rows) || cJSON_GetArraySize(rows) < 2) {
      if (rows) cJSON_Delete(rows);
      continue;
    }
    cJSON *header = cJSON_GetArrayItem(rows, 0);
    /* resolve column indices from the header row */
    int iTs = -1, iOrig = -1, iStat = -1, iMime = -1, iDig = -1;
    int hc = cJSON_GetArraySize(header);
    for (int c = 0; c < hc; c++) {
      cJSON *h = cJSON_GetArrayItem(header, c);
      const char *hs = (h && cJSON_IsString(h)) ? h->valuestring : "";
      if (!strcmp(hs, "timestamp")) iTs = c;
      else if (!strcmp(hs, "original")) iOrig = c;
      else if (!strcmp(hs, "statuscode")) iStat = c;
      else if (!strcmp(hs, "mimetype")) iMime = c;
      else if (!strcmp(hs, "digest")) iDig = c;
    }

    int rn = cJSON_GetArraySize(rows);
    for (int ri = 1; ri < rn; ri++) {
      cJSON *row = cJSON_GetArrayItem(rows, ri);
      const char *ts   = iTs   >= 0 ? cell(row, iTs)   : NULL;
      const char *orig = iOrig >= 0 ? cell(row, iOrig) : NULL;
      const char *stat = iStat >= 0 ? cell(row, iStat) : NULL;
      const char *mime = iMime >= 0 ? cell(row, iMime) : NULL;
      const char *dig  = iDig  >= 0 ? cell(row, iDig)  : NULL;

      /* uid: intelUid(SOURCE_ID, row.digest, `${tgt}|${ts}|${i}`) */
      char fb[160];
      const char *rk;
      if (dig) {
        rk = dig;
      } else {
        snprintf(fb, sizeof fb, "%s|%s|%d", host, ts ? ts : "", flat);
        rk = fb;
      }

      /* title = row.original || row._target */
      const char *title = orig ? orig : host;

      /* summary = `${tgt} · ${mime||'unknown mime'} · ${stat||'?'}` */
      char summ[256];
      snprintf(summ, sizeof summ, "%s \xc2\xb7 %s \xc2\xb7 %s",
        host, mime ? mime : "unknown mime", stat ? stat : "?");

      /* link = (ts && orig) ? https://web.archive.org/web/<ts>/<orig> : null */
      char linkbuf[1024];
      const char *link = NULL;
      if (ts && orig) {
        snprintf(linkbuf, sizeof linkbuf,
          "https://web.archive.org/web/%s/%s", ts, orig);
        link = linkbuf;
      }

      /* published_at: 14-digit ts → ISO */
      char isobuf[24];
      const char *pub = NULL;
      if (is_ts14(ts)) {
        snprintf(isobuf, sizeof isobuf,
          "%.4s-%.2s-%.2sT%.2s:%.2s:%.2sZ",
          ts, ts + 4, ts + 6, ts + 8, ts + 10, ts + 12);
        pub = isobuf;
      }

      /* properties — EXACT JS key order */
      cJSON *props = cJSON_CreateObject();
      cJSON_AddStringToObject(props, "target", host);
      cJSON_AddItemToObject(props, "timestamp",
        ts ? cJSON_CreateString(ts) : cJSON_CreateNull());
      cJSON_AddItemToObject(props, "original",
        orig ? cJSON_CreateString(orig) : cJSON_CreateNull());
      cJSON_AddItemToObject(props, "mimetype",
        mime ? cJSON_CreateString(mime) : cJSON_CreateNull());
      cJSON_AddItemToObject(props, "statuscode",
        stat ? cJSON_CreateString(stat) : cJSON_CreateNull());
      cJSON_AddItemToObject(props, "digest",
        dig ? cJSON_CreateString(dig) : cJSON_CreateNull());
      char *pj = cJSON_PrintUnformatted(props);

      char tags[96];
      snprintf(tags, sizeof tags, "[\"wayback\",\"host:%s\"]", host);

      intel_item it = {0};
      it.remote_key     = rk;
      it.title          = title;
      it.summary        = summ;
      it.link           = link;
      it.lang           = "ja";
      it.published_at   = pub;
      it.record_type    = "wayback-jp";
      it.properties_json = pj;
      it.tags_json      = tags;
      if (sink->emit(sink, &it) >= 0) n++;

      free(pj);
      cJSON_Delete(props);
      flat++;
    }
    cJSON_Delete(rows);
  }
  fprintf(stderr, "[wayback-jp] emitted %d\n", n);
  return n >= 0 ? 0 : -1;
}

static const source_def wayback_jp_def = {
  .id = "wayback-jp", .collector = "cyber",
  .name = "Wayback CDX (JP gov)", .name_ja = "Wayback Machine 日本政府",
   .update_interval_sec = 86400, .run = run,
};
REGISTER_SOURCE(wayback_jp_def)
